Lib/shutil.py

```diff
@@ -72,14 +83,124 @@ class RegistryError(Exception):
+    """Copy data from one regular mmap-like fd to another by using
+    high-performance sendfile(2) syscall.
+    This should work on Linux >= 2.6.33 and Solaris only.
+    """
+    # Note: copyfileobj() is left alone in order to not introduce any
+    # unexpected breakage. Possible risks by using zero-copy calls
+    # in copyfileobj() are:
+    # - fdst cannot be open in "a"(ppend) mode
+    # - fsrc and fdst may be open in "t"(ext) mode
+    # - fsrc may be a BufferedReader (which hides unread data in a buffer),
+    #   GzipFile (which decompresses data), HTTPResponse (which decodes
+    #   chunks).
+    # - possibly others (e.g. encrypted fs/partition?)
+    global _HAS_SENDFILE
+    try:
+        infd = fsrc.fileno()
+        outfd = fdst.fileno()
+    except Exception as err:
+        raise _GiveupOnFastCopy(err)  # not a regular file
+
+    # Hopefully the whole file will be copied in a single call.
+    # sendfile() is called in a loop 'till EOF is reached (0 return)
+    # so a bufsize smaller or bigger than the actual file size
+    # should not make any difference, also in case the file content
+    # changes while being copied.
+    try:
+        blocksize = max(os.fstat(infd).st_size, 2 ** 23)  # min 8MB
+    except Exception:
+        blocksize = 2 ** 27  # 128MB
+
+    offset = 0
+    while True:
+        try:
+            sent = os.sendfile(outfd, infd, offset, blocksize)
+        except OSError as err:
+            # ...in oder to have a more informative exception.
+            err.filename = fsrc.name
+            err.filename2 = fdst.name
+
+            if err.errno == errno.ENOTSOCK:
+                # sendfile() on this platform (probably Linux < 2.6.33)
+                # does not support copies between regular files (only
+                # sockets).
+                _HAS_SENDFILE = False
+                raise _GiveupOnFastCopy(err)
+
+            if err.errno == errno.ENOSPC:  # filesystem is full
+                raise err from None
+
+            # Give up on first call and if no data was copied.
+            if offset == 0 and os.lseek(outfd, 0, os.SEEK_CUR) == 0:
+                raise _GiveupOnFastCopy(err)
+
+            raise err
+        else:
+            if sent == 0:
+                break  # EOF
+            offset += sent
+
+def _copybinfileobj(fsrc, fdst, length=COPY_BUFSIZE):
+    """Copy 2 regular file objects open in binary mode."""
+    # Localize variable access to minimize overhead.
+    fsrc_readinto = fsrc.readinto
+    fdst_write = fdst.write
+    with memoryview(bytearray(length)) as mv:
+        while True:
+            n = fsrc_readinto(mv)
+            if not n:
+                break
+            elif n < length:
+                fdst_write(mv[:n])
+            else:
+                fdst_write(mv)
+
+def _is_binary_files_pair(fsrc, fdst):
+    return hasattr(fsrc, 'readinto') and \
+        isinstance(fsrc, io.BytesIO) or 'b' in getattr(fsrc, 'mode', '') and \
```

Comment:
Which objects provide `readinto` in text mode? Is it worth the function call and extra tests rather than handling `AttributeError` (no `readinto`) and `TypeError` (can't write bytes) with an inline try-except?

Reply:
I think catching `TypeError` on `fdst.write()` is too risky as we can deal with any kind of custom file-like object being passed here. It must be noted that the extra cost of this function is payed by users of `copyfileobj()` only (e.g. `tarfile` and `zipfile` modules). `copyfile()` (and others) will skip this check and call `_copybinfileobj()` directly.

Reply:
I benchmarked `_copybinfileobj()` (I hadn't yet) and it turns out it's only slightly faster for 512MB files but considerably slower for 8MB and 128KB files so I am gonna remove it.

Reply:
I don't know how you're testing, but the performance difference with `readinto` depends on whether the source file is already in the system cache. Otherwise, of course memory operations will be dwarfed by considerably slower disk I/O.

Reply:
This is how I'm testing it:

```bash
$ python -c "import os; f = open('f1', 'wb'); f.write(os.urandom(8 * 1024 * 1024))"
$ time ./python -m timeit -s 'import shutil; p1 = "f1"; p2 = "f2"' 'shutil.copyfile(p1, p2)'
```

Reply:
Sorry, wrong cmdline. This is the correct one:

```bash
$ python -c "import os; f = open('f1', 'wb'); f.write(os.urandom(8 * 1024 * 1024))"
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
```

Original:

```
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
200000 loops, best of 5: 1.36 usec per loop
```
Memoryview / readinto:
```
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
100000 loops, best of 5: 3.22 usec per loop
```

Reply:
Does the original case use a 1 MiB `length` for the 8 MiB file, or is that the original 16 KiB length?

Reply:
Both cases use 1MiB.

Reply:
I checked Linux with the default 16 KiB `length`. Using `readinto` provides no benefit on larger files and degrades performance on smaller files. Given only a modest 10% increase in Windows with large files, I think your decision to remove it is the right call. It's not worth the bookkeeping overhead to keep two versions, and it's not clear even on Windows where a good dividing line would be if we `[f]stat` the source file to check `st_size`.

Reply:
I wrote a batch script to figure out timings on Windows more easily and this is the result (first value is original `copyfileobj()` implementation, second value is the `memoryview()` variant):

```
8MB file
1000 loops, best of 5: 343 usec per loop
500 loops, best of 5: 478 usec per loop

64MB file
500 loops, best of 5: 474 usec per loop
500 loops, best of 5: 554 usec per loop

128MB file
200 loops, best of 5: 1.06 msec per loop
500 loops, best of 5: 640 usec per loop

256MB file
1 loop, best of 5: 286 msec per loop
5 loops, best of 5: 36.1 msec per loop

512MB file
1 loop, best of 5: 293 msec per loop
5 loops, best of 5: 36.7 msec per loop
```

I think the `memoryview()` variant after 128MB is so much faster that it is worth to have the dual implementation and use it **from `copyfile()` function only** if on Windows and file size > 128MB. I will do it in my other PR/branch.

On the other hand, the same test on Linux shows there is no relevant difference.

Reply:
Done: https://github.com/python/cpython/pull/7681/files

Reply:
This command is wrong:

    ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"

The files need to be opened for each pass of the loop, not the setup. That explains the unexpected results. I corrected it to open the files in the loop statement instead of the setup and tested a broad range of file sizes. In the table below all times are best of five for the give number of loops, and normalized overall to make the 64KiB result in the RI_S column equal to 100 time units. I discuss the RI_S case in more detail below.

```
   SIZE | LOOPS |   R_16K ||    R_1M   %CHNG |  RI_1M   %CHNG |  RI_S   %CHNG
--------+-------+---------||-----------------+----------------+--------------
  1 GiB |     5 | 1060870 || 1003478    -5.4 | 977391    -7.9 | 963478   -9.2
512 MiB |    10 |  323478 ||  283478   -12.4 | 217391   -32.8 | 213913  -33.9
256 MiB |    20 |  163304 ||  146435   -10.3 | 112870   -30.9 | 110957  -32.1
128 MiB |    40 |   80174 ||   74609    -6.9 |  55478   -30.8 |  55478  -30.8
 64 MiB |    80 |   39652 ||   34783   -12.3 |  27304   -31.1 |  27130  -31.6
 32 MiB |   160 |   19478 ||   17913    -8.0 |  13809   -29.1 |  13478  -30.8
 16 MiB |   320 |    9739 ||    8887    -8.7 |   6887   -29.3 |   6835  -29.8
  8 MiB |   640 |    4904 ||    4713    -3.9 |   3652   -25.5 |   3548  -27.7
  4 MiB |  1280 |    2504 ||    2348    -6.2 |   1948   -22.2 |   1913  -23.6
  2 MiB |  2560 |    1150 ||    1193     3.7 |   1002   -12.9 |    991  -13.8
  1 MiB |  5120 |     649 ||     697     7.4 |    659     1.5 |    663    2.2
512 KiB | 10240 |     388 ||     553    42.5 |    499    28.6 |    322  -17.0
256 KiB | 20480 |     245 ||     459    87.3 |    410    67.3 |    205  -16.3
128 KiB | 30720 |     170 ||     388   128.2 |    362   112.9 |    139  -18.2
 64 KiB | 40960 |     123 ||     353   187.0 |    343   178.9 |    100  -18.7

    R_16K -- read 16 KiB
    R_1M  -- read 1 MiB
    RI_1M -- readinto 1 MiB
    RI_S  -- readinto source size up to 1 MiB
```

Originally I had tested at 128 MiB with a custom test script to focus on the effects of cached vs non-cached I/O. I assumed the results would be similar for other cases. As shown in the RI_1M column, that's basically true for files larger than 1 megabyte. But there's a significant performance degradation for smaller files.  In the RI_S case, I address this by calling `os.fstat` on the source file to cap the `length` of the `bytearray` at its size. This avoids wastefully over-allocating a zeroed `byterray`. 

RI_S also experiments with calling `SetInformationByHandle` : `FileEndOfFileInfo` (prototyped using ctypes) to avoid having to repeatedly extend the file when `length` is less than the size of the source file. (Note that Python's `truncate` method is of no use here since it zeros the file.) `CopyFileEx` does this, so I figured it was worth a try. This does provide a modest performance increase. (Note that I mistakenly included the `length == size` boundary, which is apparent in the 1 MiB trial.) I don't know if it's significant enough to justify implementing `_winapi.SetFileInformationByHandle`. 

If I have time, I may run another experiment using `mmap` to read into a sliding window of the destination file. It already implements setting the end of the file, albeit with the less efficient combination of `SetFilePointer` and `SetEndOfFile`. (This way requires 4 system calls instead of 1 to set the end of the file.)

Below is the code for RI_S:

     import os

    _WINDOWS = (os.name == 'nt')

    if _WINDOWS:
        import msvcrt
        import ctypes
        from ctypes import wintypes
        
        COPY_BUFSIZE = 1024 *1024
        
        kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)

        FileEndOfFileInfo = 6

        FILE_INFO_BY_HANDLE_CLASS = wintypes.DWORD
        kernel32.SetFileInformationByHandle.argtypes = (
            wintypes.HANDLE,           # _In_ hFile
            FILE_INFO_BY_HANDLE_CLASS, # _In_ FileInformationClass
            wintypes.LPVOID,           # _In_ lpFileInformation
            wintypes.DWORD)            # _In_ dwBufferSize

    def copyfileobj(fsrc, fdst, length=COPY_BUFSIZE):
        size = os.fstat(fsrc.fileno()).st_size
        if length > size:
            length = size
        elif _WINDOWS:
            info = wintypes.LARGE_INTEGER(size)
            if not kernel32.SetFileInformationByHandle(
                        msvcrt.get_osfhandle(fdst.fileno()), FileEndOfFileInfo, 
                        ctypes.byref(info), ctypes.sizeof(info)):
                raise ctypes.WinError(ctypes.get_last_error())
        fsrc_readinto = fsrc.readinto
        fdst_write = fdst.write
        with memoryview(bytearray(length)) as mv:
            while True:
                n = fsrc_readinto(mv)
                if not n:
                    break
                elif n < length:
                    with mv[:n] as smv:
                        fdst_write(smv)
                else:
                    fdst_write(mv)

Reply:
Interesting. Thanks for the very detailed benchmark. I updated the other branch which now dynamically sets `memoryview()` size based on file size (https://github.com/python/cpython/pull/7681/commits/93ebc1ffdc547f4057a164922c7962186ef1c320) and I confirm using the `readinto()` variant is faster also for smaller files.

Reply:
https://github.com/python/cpython/pull/7681 was merged